compiler/noirc_frontend/src/ownership/last_uses.rs

```diff
@@ -0,0 +1,410 @@
+    ///   one use but can be multiple if the last use is spread across several `if` or `match`
+    ///   branches. E.g:
+    ///   ```noir
+    ///   fn bar() {
+    ///       let x = 2;
+    ///       if true {
+    ///           println(x);
+    ///       } else {
+    ///           assert(x < 5);
+    ///       }
+    ///   }
+    ///   ```
+    ///   `x` above has two last uses, one in each if branch.
+    last_uses: HashMap<LocalId, (/*loop index*/ usize, Branches)>,
+}
+
+impl Context {
+    /// Traverse the given function and return the last use(s) of each local variable.
+    /// A variable may have multiple last uses if it was last used within a conditional expression.
+    pub(super) fn find_last_uses_of_variables(
+        function: &Function,
+    ) -> HashMap<LocalId, Vec<IdentId>> {
+        let mut context =
+            LastUseContext { current_loop_and_branch: Vec::new(), last_uses: HashMap::default() };
+        context.push_loop_scope();
+        for (parameter, ..) in &function.parameters {
+            context.declare_variable(*parameter);
+        }
+        context.track_variables_in_expression(&function.body);
+        context.get_variables_to_move()
+    }
+}
+
+impl LastUseContext {
+    fn push_loop_scope(&mut self) {
+        self.current_loop_and_branch.push(BranchesPath::Here);
+    }
+
+    fn pop_loop_scope(&mut self) {
+        self.current_loop_and_branch.pop();
+    }
+
+    fn branch(&mut self, branch_index: usize) {
+        let path =
+            self.current_loop_and_branch.last_mut().expect("We should always have at least 1 path");
+        let rest = Box::new(std::mem::replace(path, BranchesPath::Here));
+        *path = BranchesPath::IfOrMatch { branch_index, rest };
+    }
+
+    fn unbranch(&mut self) {
+        let path =
+            self.current_loop_and_branch.last_mut().expect("We should always have at least 1 path");
+        let rest = std::mem::replace(path, BranchesPath::Here);
+
+        match rest {
+            BranchesPath::Here => panic!("unbranch called without any branches"),
+            BranchesPath::IfOrMatch { branch_index: _, rest } => *path = *rest,
+        }
+    }
+
+    fn loop_index(&self) -> usize {
+        self.current_loop_and_branch.len() - 1
+    }
+
+    fn declare_variable(&mut self, id: LocalId) {
+        let loop_index = self.current_loop_and_branch.len() - 1;
+        self.last_uses.insert(id, (loop_index, Branches::None));
+    }
+
+    fn remember_use_of_variable(&mut self, id: LocalId, variable: IdentId) {
+        let path =
+            self.current_loop_and_branch.last().expect("We should always have at least 1 scope");
+        let loop_index = self.loop_index();
+
+        if let Some((variable_loop_index, uses)) = self.last_uses.get_mut(&id) {
+            if *variable_loop_index == loop_index {
+                Self::remember_use_of_variable_rec(uses, path, variable);
+            } else {
+                *uses = Branches::None;
+            }
```

Comment:
Is there a way to relate these two if branches to codes that would trigger them? I think I understand that if the variable is in a loop index and we are currently in that index, we update the use. But if it's in a different index (an upper one?) then why this is set to None?

Reply:
There's code examples in the doc comments in the fields of the context struct for this case but I'll add a comment to mention it here.

This is set to None if we're in an upper loop index since that means we are in a loop that the variable wasn't defined in, and a variable can never be moved into a loop (it needs to be cloned). This is a situation like:
```
let foo = bar;  // loop index 0
loop {
    println(foo);  // loop index 1, needs to be copied
}
```
If it was defined in the same loop, it can be moved:
```
loop {
    let foo = bar; // loop index 1
    println(foo);  // loop index 1, can be moved
}
```